text/0000-unsized-rvalues.md

```diff
@@ -0,0 +1,167 @@
+
+One particularly common example is passing closures that consume their environment without using monomorphization. One would like for this code to work:
+
+```Rust
+fn takes_closure(f: FnOnce()) { f(); }
+```
+
+But today you have to use a hack, such as taking a `Box<FnBox<()>>`.
+
+2) Allocating a runtime-sized variable on the stack is important for good performance in some use-cases - see RFC #1808, which this is intended to supersede.
+
+# Detailed design
+[design]: #detailed-design
+
+## Unsized Rvalues - language
+
+Remove the rule that requires all locals and rvalues to have a sized type. Instead, require the following:
+a) The following expressions must always return a Sized type:
+    a1) Function calls, method calls, operator expressions
+        - implementing unsized return values for function calls would require the *called function* to do the alloca in our stack frame.
+    a2) ADT expressions
+        - see alternatives
+    a3) cast expressions
+        - this seems like an implementation simplicity thing. These can only be trivial casts.
+b) The RHS of assignment expressions must always have a Sized type.
+    - Assigning an unsized type is impossible because we don't know how much memory is available at the destination. This applies to ExprAssign assignments and not to StmtLet let-statements.
+
+This also allows passing unsized values to functions, with the ABI being as if a `&move` pointer was passed (a `(by-move-data, extra)` pair). This also means that methods taking `self` by value are object-safe, though vtable shims are sometimes needed to translate the ABI (as the callee-side intentionally does not pass `extra` to the fn in the vtable, no vtable shim is needed if the vtable function already takes its argument indirectly).
+
+For example:
+
+```Rust
+struct StringData {
+    len: usize,
+    data: [u8],
+}
+
+fn foo(s1: Box<StringData>, s2: Box<StringData>, cond: bool) {
+    // this creates a VLA copy of either `s1.1` or `s2.1` on
+    // the stack.
+    let mut s = if cond {
+        s1.1
+    } else {
+        s2.1
+    };
+    drop(s1);
+    drop(s2);
+    foo(s);
+}
+
+fn example(f: for<'a> FnOnce(&'a X<'a>)) {
+    let x = X::new();
+    f(x); // aka FnOnce::call_once(f, (x,));
+}
+```
+
+## VLA expressions
+
+Allow repeat expressions to capture variables from their surrounding environment. If a repeat expression captures such a variable, it has type `[T]` with the length being evaluated at run-time. If the repeat expression does not capture any variable, the length is evaluated at compile-time. For example:
+```Rust
+extern "C" {
+   fn random() -> usize;
+}
+
+fn foo(n: usize) {
+    let x = [0u8; n]; // x: [u8]
+    let x = [0u8; n + (random() % 100)]; // x: [u8]
+    let x = [0u8; 42]; // x: [u8; 42], like today
+    let x = [0u8; random() % 100]; //~ ERROR constant evaluation error
```

Comment:
Context: I *already* have a branch where the MIR const qualifier runs before the old `rustc_const_eval` ever looks at the expression, *and* we already have to evaluate it anyway to get the type, so, the rule could be:
Can the repeat count be evaluated as a constant? If so, create a `[T; N]` type, otherwise, a `[T]`.

The implementation problem is not hard, not unreasonably so, and it's actually *solved*.

But it's not *the* problem being solved by such an almost syntactical approach, **backwards compatibility** is. If we start making more things const-evaluatable, the user can have a `[T]` that becomes `[T; const]` in a newer version, which could cause type errors.